### ISSUE TITLE
[Feature Request] Do not hardcode save file extension (and allow changing in configuration file)

### ISSUE THREAD

**Issue by sofakng** [MiSTer-devel/Main_MiSTer#754]
MiSTer currently defaults all save file extensions to '.sav' without any options to change it:

https://github.com/MiSTer-devel/Main_MiSTer/blob/b5ea246ffe003dc0e15bb32e751f3f5292d7ff9f/file_io.cpp#L876

I'd like to request a configuration setting (MiSTer.ini) to allow changing the default extension to any value.

This would eliminate the need to rename .sav to .srm when copying (or sharing) with flash cartridges, emulators, etc.  It would also help in situations where the save games are shared on a network drive (for example) and multiple devices/programs/emulators are sharing the file but expect a specific file extension.

I'd be willing to create a pull request but I'd like feedback before I put in the effort.

**Comment by sofakng** [comment]
I agree that if you are copying the files then it's easy to rename the files during the copy, but my situation has multiple devices on my network sharing the same files.  For example, using an emulator on my desktop PC and MiSTer on my TV/CRT.

I could be mistaken (and probably am) but it looks like '.srm' is more common than '.sav' for existing flash cartridges (Everdrive, etc) and emulators (mGBA, Retroarch, etc).

Regardless, I think it would be useful to change the extension or I perhaps even changing the default to '.srm' but obviously this would break existing setups so that probably isn't a good idea.

**Comment by birdybro** [comment]
Yeah .srm is more common, but it's naming is also a misnomer since it's short for "Saved RAM", but there are more than just battery backed RAM for saves across retro consoles, so .srm doesn't make much sense nowadays. EEPROM for instance aren't RAM. Some systems like the PSX saved to a memory card externally that wasn't RAM either, so .srm was mostly used because it just had been used for a very long time.

**Comment by sofakng** [comment]
OK - That's a good argument for not changing the default extension but I'm hoping that adding this as a configuration item wouldn't be too confusing and still prove useful to several users.

**Comment by sorgelig** [comment]
save files are not exactly compatible with other devices or emulators. It also depends on specific core. So better not to confuse or give misunderstanding by mimicking file extensions.

**Comment by sofakng** [comment]
Thanks for the link.  However, it looks like several of cores/emulators are already compatible with the files except for the file extension.

For example, I tested the SNES core and the saves are directly compatible with bsnes emulator.  The same for NES core and Mesen emulator and GBA core and mGBA.  I haven't  tested any others, but at least for these the issue isn't the files not being compatible but just that the extensions aren't what is expected.

**Comment by birdybro** [comment]
That isn't true for all cores. 

For instance, the MegaCD core concatenates the internal backup memory with the 512kb cartridge memory if you have the external cart memory option turned on. The Genesis core initializes saves to 0xFF, but some older emulators may initialize to 0x00, so padding is important upon size conversion, which sometimes they have to be size converted.

https://github.com/euan-forrester/save-file-converter/tree/main/frontend/src/save-formats/Mister

You can see the long back and forth here (and it's spread into issues in a few other mister core repos as well) --> https://github.com/euan-forrester/save-file-converter/issues/80

More importantly, some emulators and the Everdrive do writes and reads on genesis save files as 16-bit words, but MiSTer's core reads them as 8-bit words (which I believe more modern emulators do now, like gpgx). Here's the mister <-> everdrive conversion tool that uses the code that Kitrinx came up with to solve that problem --> https://github.com/superg/srmtools - which will do a conversion between the two.

But mostly it's just the Genesis and MegaCD, the edge cases of different sizes and different padding haven't shown effects on other cores yet, but euan-forrester's tool still does the job right.

**Comment by sofakng** [comment]
You guys have definitely provided good arguments and I can understand the reasoning for not switching the default extension but I don't see the harm in allowing a user option (with core overrides)?

The default would (and should) stay as '.sav' but if somebody wants to knowingly override it then I don't see the harm?
